### PR TITLE
fix: increment gas on retry if out of gas

### PR DESCRIPTION
**Pull request by Abhikumar98**
# FEATURE

Retry with increased gas if transaction fails with out of gas error [LF-2862]

## TL;DR
Automatically retry transactions with increased gas if they fail with out of gas errors.

## DESCRIPTION
This feature automatically retries transactions with increased gas if they fail with out of gas errors. When attempting to send a transaction for bridging or swapping, if the initial gas is too low and the transaction fails with an out of gas error, the feature will automatically retry the transaction with increased gas to ensure that it is successfully processed.

## STEPS TO TEST
1. Reduce the gas to 21000 from Metamask before sending a transaction for bridging or swapping.
2. Observe that the transaction fails with an out of gas error.
3. The feature will automatically retry the transaction with increased gas.
4. Verify that the transaction is successfully processed.

## EXPECTED RESULTS
The feature should automatically retry transactions with increased gas if they fail with out of gas errors, and the transactions should be successfully processed.

## CHECKLIST
- [ ] I have performed a self-review of my code.
- [ ] I have commented the code I wrote to make sure the next developers can understand it.
- [ ] I have added tests that cover the new functionality.
- [ ] I have made sure that existing tests are still passing.

https://user-images.githubusercontent.com/35174288/236535777-e4a67fc2-ad13-4d41-9547-3fcd46431f22.mp4

[LF-2862]: https://lifi.atlassian.net/browse/LF-2862?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/utils/preRestart.ts [hunk 1]
```diff
@@ -1,28 +1,38 @@
...
   )
   const isGasPriceError = route.steps[index].execution?.process.some(
     (p) => p.error?.code === LifiErrorCode.TransactionUnderpriced
   )
 
+  const { transactionRequest } = route.steps[index]
+
   if (isGasLimitError) {
+    if (transactionRequest) {
+      const gasLimit = signer.estimateGas(transactionRequest)
```
Comment (chybisov) [inline src/utils/preRestart.ts:30]: Can this potentially throw something?

Comment (Abhikumar98) [reply src/utils/preRestart.ts:30]: This ideally shouldn't throw any errors until there's an issue from RPC or an incorrect `transactionRequest` being sent to it.

But to be on safer side, wrapped in `try...catch` block and also handling the promises correctly now.

File: src/LiFi.unit.spec.ts [hunk 2]
```diff
@@ -1,28 +1,86 @@
...
+          gasAmount: '123',
+          gasAmountUSD: '123',
+          gasPrice: '123',
+          gasToken: '123',
+          gasUsed: '123',
+        },
+      })
+    ),
+  },
+}))
```
Comment (chybisov) [inline src/LiFi.unit.spec.ts:39]: It looks okay, but can we please use `msw` here, like in the rest of the tests, to mock responses from our BE? The same applies to all tests, if any.

File: src/LiFi.unit.spec.ts [hunk 3]
```diff
@@ -1,28 +1,86 @@
...
+
+vi.mock('./allowance/utils', () => ({
+  getApproved: vi.fn(() => Promise.resolve([])),
+}))
+
+// vi.mock('./execution/utils', () => ({
+// waitForReceivingTransaction: vi.fn(() => Promise.resolve()),
+// getProcessMessage: vi.fn(() => 'message'),
+// checkStepSlippageThreshold: vi.fn(() => true),
+// }))
```
Comment (chybisov) [inline src/LiFi.unit.spec.ts:60]: Please remove unnecessary comments when you finish with the PR.

File: test/fixtures.ts [hunk 4]
```diff
@@ -168,3 +168,97 @@ export const buildRouteObject = ({
     state: 'NOT_INSURABLE',
   },
 })
+
+export const mockChainsResponse = [
```
Comment (chybisov) [inline test/fixtures.ts:172]: This should act as `msw` response.

**Review by chybisov — CHANGES_REQUESTED** [review]

File: src/LiFi.unit.spec.ts (hunk 2 above)
Comment (Abhikumar98) [reply src/LiFi.unit.spec.ts:39]: Sure, I'll make this change

File: test/fixtures.ts (hunk 4 above)
Comment (Abhikumar98) [reply test/fixtures.ts:172]: alright, making this change!

File: src/LiFi.unit.spec.ts [hunk 5]
```diff
@@ -211,4 +269,49 @@ describe('LIFI SDK', () => {
...
+    it('should pick up gas estimation from signer', async () => {
+      const route = buildRouteObject({
+        step,
+      })
+
+      await lifi.executeRoute(signer, route)
+
+      expect(signer.sendTransaction).toHaveBeenCalledWith({
+        gasLimit: '125000',
+        // TODO: Check the cause for gasLimit being outside transactionRequest. Currently working as expected in widget
```
Comment (Addminus) [inline src/LiFi.unit.spec.ts:283]: Do you want to fix that in this PR or in another one?

If you want to leave that as TODO here, please mention it in the PR description

**Review by Addminus — CHANGES_REQUESTED** [review]

File: src/utils/preRestart.ts [hunk 6]
```diff
@@ -1,28 +1,41 @@
...
 
+  const { transactionRequest } = route.steps[index]
+
   if (isGasLimitError) {
+    if (transactionRequest) {
+      let gasLimit = transactionRequest.gasLimit
+
+      try {
+        gasLimit = await signer.estimateGas(transactionRequest)
+      } catch (error) {}
```
[the remark this answers is not in the document]
Comment (Abhikumar98) [reply src/utils/preRestart.ts:34]: I recently had a conversation with @chybisov where he mentioned that this is best being ignored.

although, at that time I'd mistakenly written `console.log` instead of `console.error`.

I'd still prefer having an error being logged. because the rest of the error are being handled in custom error handlers.

wdyt?
cc: @chybisov

[the remark this answers is not in the document]
Comment (chybisov) [reply src/utils/preRestart.ts:34]: We should have a fallback in case it throws, but we don't log anything in the SDK now, so I see no reason to log this one for now, but maybe we should come up with something in the future. We should especially avoid `console.error` as it can annoy some envs.

File: test/fixtures.ts (hunk 4 above)
Comment (chybisov) [reply test/fixtures.ts:172]: @Abhikumar98 please don't resolve conversations until fixed 😄

Comment (Abhikumar98) [reply test/fixtures.ts:172]: Now the mockResponse is being used in the msw mocks. Does this need to be relocated to another file?

File: src/LiFi.unit.spec.ts (hunk 5 above)
Comment (Abhikumar98) [reply src/LiFi.unit.spec.ts:283]: Duplicate. Resolving this

**Review by chybisov — CHANGES_REQUESTED** [review]

File: src/LiFi.unit.handlers.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      )
+  ),
+  rest.get(`${config.apiUrl}/chains`, async (_, response, context) =>
+    response(
+      context.status(200),
+      context.json({
+        chains: mockChainsResponse,
+      })
+    )
+  ),
```
Comment (chybisov) [inline src/LiFi.unit.handlers.ts:34]: Couldn't find where we use this endpoint in tests. 😅

**Review by chybisov — CHANGES_REQUESTED** [review]

File: yarn.lock [hunk 8]
```diff
@@ -6602,6 +6603,15 @@ __metadata:
...
 
+"tsc@npm:^2.0.4":
+  version: 2.0.4
+  resolution: "tsc@npm:2.0.4"
+  bin:
+    tsc: bin/tsc
+  checksum: 50b10240887424c66454687f0fca4e319713b35cb44f0dab228719eb4dd757251f83c3d79e60d29dfbe90c2d6f1272f793b83e75a9fca2c622b4213ad3eb8cb1
+  languageName: node
+  linkType: hard
+
```
[the remark this answers is not in the document]
Comment (Abhikumar98) [reply yarn.lock:6614]: I was testing with tsc. I'll remove this.
Got pushed by mistake

File: src/LiFi.unit.handlers.ts (hunk 7 above)
Comment (Abhikumar98) [reply src/LiFi.unit.handlers.ts:34]: This is being called from inside lifi.execute

File: src/LiFi.unit.spec.ts [hunk 9]
```diff
@@ -210,4 +259,28 @@ describe('LIFI SDK', () => {
...
+          data: '0xdata',
+          from: '0x552008c0f6870c2f77e5cC1d2eb9bdff03e30Ea0',
+          gasLimit: '682701',
+          gasPrice: '0x27c01c1727',
+          to: '0x1231DEB6f5749EF6cE6943a275A1D3E7486F4EaE',
+          value: '0x0600830dbc7f5bf7',
+        },
+      })
+    })
+  })
```
Comment (chybisov) [inline src/LiFi.unit.spec.ts:285]: @Abhikumar98 I just realized that we test functionality for `RouteExecutionManager` here and LIFI class just extends it. Does it make sense to move these tests to RouteExecutionManager.unit.spec.ts?

File: src/LiFi.unit.handlers.ts (hunk 7 above)
Comment (chybisov) [reply src/LiFi.unit.handlers.ts:34]: > The api endpoints for this is being called within the lifi.execute method. Hence mocking the response for that over here.

I can't seem to find that place at the first glance. Please share 🙏😄

**Review by chybisov — CHANGES_REQUESTED** [review]

File: src/LiFi.unit.spec.ts (hunk 9 above)
Comment (Abhikumar98) [reply src/LiFi.unit.spec.ts:285]: Yes we can do that. That'd would be better

File: src/LiFi.unit.handlers.ts (hunk 7 above)
Comment (Abhikumar98) [reply src/LiFi.unit.handlers.ts]: This is being called while creating new instance of `StepExecutionManager`.

Inside the execute method, we are using `ChainService.getInstance()`, which fetches all the chains from the Backend.

File: src/utils/preRestart.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Addminus) [reply src/utils/preRestart.ts:34]: alright! We should make sure that a failure here does not break the process. We should see that estimation as an addition. If we don't set anything, metamask sets its own estimations.
So it's sort of a nice to have if it works but not breaking and falling back to metamask defaults if it does not work. It looks like the code does exactly that.

[the remark this answers is not in the document]
Comment (Addminus) [reply src/utils/preRestart.ts:34]: jup true

File: src/execution/StepExecutionManager.ts [hunk 10]
```diff
@@ -153,6 +153,14 @@ export class StepExecutionManager {
...
               customConfig.maxPriorityFeePerGas
             transactionRequest.maxFeePerGas = customConfig.maxFeePerGas
+          } else {
+            const estimatedGasLimit = await signer.estimateGas(
+              transactionRequest
+            )
+
+            transactionRequest.gasLimit = `${Math.round(
+              Number(estimatedGasLimit) * 1.25
+            )}`
```
Comment (chybisov) [inline src/execution/StepExecutionManager.ts:163]: Reviewing this once again, I realized that the return type from `estimateGas` is `BigNumber`, but we convert it to a plain `Number` here, and sure, it's an edge case for gas estimation, but this potentially can lead to an error if we get a big number that exceeds `Number.MAX_SAFE_INTEGER`. I think we should switch to `BigInt` here.

Also, when we switch to `viem`, this function starts to return `BigInt`, which just reinforces my words 🙂

**Review by chybisov — CHANGES_REQUESTED** [review]

File: src/LiFi.unit.handlers.ts (hunk 7 above)
Comment (chybisov) [reply src/LiFi.unit.handlers.ts]: @Abhikumar98 do we need that huge mock here, or can we just mock something simple just to avoid a call to BE?

File: src/execution/StepExecutionManager.ts (hunk 10 above)
Comment (chybisov) [reply src/execution/StepExecutionManager.ts:163]: `${BigInt(estimatedGasLimit) * 125n / 100n}`

File: src/LiFi.unit.handlers.ts (hunk 7 above)
Comment (Abhikumar98) [reply src/LiFi.unit.handlers.ts]: We'll need Polygon, because that's where the swap is happening. I'll remove the other one

File: src/execution/StepExecutionManager.ts (hunk 10 above)
Comment (Abhikumar98) [reply src/execution/StepExecutionManager.ts]: Hey @chybisov,
Pushed the changes for this :)

_`/prerestart.ts`_
CleanShot 2023-05-17 at 16 47 20

_`StepExecutionManager.ts`_
CleanShot 2023-05-17 at 16 46 50

File: src/utils/preRestart.ts [hunk 11]
```diff
@@ -1,28 +1,45 @@
...
+
+      if (gasLimit) {
+        transactionRequest.gasLimit = `${
+          (BigInt(gasLimit?.toString()) * 125n) / 100n
```
Comment (chybisov) [inline src/utils/preRestart.ts:38]: Unnecessary optional chaining here due to the above check. Also, passing undefined to BigInt will throw TypeError.

File: src/execution/StepExecutionManager.ts [hunk 12]
```diff
@@ -153,6 +153,16 @@ export class StepExecutionManager {
...
               customConfig.maxPriorityFeePerGas
             transactionRequest.maxFeePerGas = customConfig.maxFeePerGas
+          } else {
+            const estimatedGasLimit = await signer.estimateGas(
+              transactionRequest
+            )
+
+            if (estimatedGasLimit) {
+              transactionRequest.gasLimit = `${
+                (BigInt(estimatedGasLimit?.toString()) * 125n) / 100n
```
Comment (chybisov) [inline src/execution/StepExecutionManager.ts:163]: Same unnecessary optional chaining.

**Review by chybisov — CHANGES_REQUESTED** [review]

**Review by chybisov — APPROVED** [review]
